Add an eslint rule to use cross-environment SVG primitives

youknowriad opened the pull request:
Still WIP, the idea is to try to force the usage of the SVG primitives instead of the raw components to ensure mobile compatibility.

 - I don't like that this is forcing us to add a dependency: `blocks -> components`. Makes me think these primitives should be better in the `element` package.
 - The eslint rule is too aggressive at the moment, it also catches props named "path"

youknowriad committed:
Add an eslint rule to use cross-environment SVG primitives

hypest commented on .eslintrc.js:
```diff
@@ -113,6 +113,10 @@ module.exports = {
 				message: 'Deprecated functions must be removed before releasing this version.',
 			},
+			{
+				selector: 'JSXIdentifier[name=/^(path|svg)$/]',
```
Should we also add `<g>` and `<polygon>` that we already have componentized as well?

gziolo replied:
Yes, sure. 

> The eslint rule is too aggressive at the moment, it also catches props named "path"

As Riad mentioned, this rule needs to be updated anyways.

gziolo commented:
Many thanks for opening this PR @youknowriad ❤️

gziolo commented:
> I don't like that this is forcing us to add a dependency: blocks -> components. Makes me think these primitives should be better in the element package.

```js
<SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><Path d="M19 7h-1V5h-4v2h-4V5H6v2H5c-1.1 0-2 .9-2 2v10h18V9c0-1.1-.9-2-2-2zm0 10H5V9h14v8z" /></SVG> 
```

I personally think that `@wordpress/blocks` shouldn't be so opinionated about the default icon. If we could register a default icon instead it would solve the issue.

gziolo commented:
@Tug if you have some time, you can help to land this one.

gziolo committed:
Extend the list of forbidden elements
Uses an existing react eslint rule to perform validation.
Adds new components to make the existing icons mobile app friendly.

gziolo commented:
This should be ready to ship. I figured out that there is an existing rule in react linting library which we could take advantage of: https://github.com/yannickcr/eslint-plugin-react/blob/HEAD/docs/rules/forbid-elements.md.

aduth reviewed (dismissed):
This white-screens for me (and apparently in Travis) with the following error:

```
Uncaught TypeError: Cannot read property 'SVG' of undefined
```

Maybe missing a dependency? (From `blocks` to `components`)

gziolo commented:
> Maybe missing a dependency? (From `blocks` to `components`)

Yes, it is an actual issue which @youknowriad raised earlier. I will refactor code to avoid using SVG inside `@wordpress/blocks` package. It should be UI agnostic.

gziolo committed:
Change the way the default icon for block is set

gziolo commented on packages/editor/src/components/block-icon/index.js:
```diff
@@ -6,19 +6,24 @@ import classnames from 'classnames';
 
 function renderIcon( icon ) {
 	if ( 'string' === typeof icon ) {
+		if ( icon === 'block-default' ) {
```
@jasmussen - is it okay to handle SVG replacement in here?

jasmussen replied:
Can you elaborate a little bit? Not sure I understand the context.

gziolo replied:
See also my previous comment: https://github.com/WordPress/gutenberg/pull/10652#issuecomment-430682576.

gziolo replied:
I changed my mind when entertaining with the exisitng implementation. I reverted all change in `normalizeIconObject` to use `default-block` dashicon. However, it is replaced with the new nice `SVG` icon inside `BlockEdit`. This should behave exactly the same in the UI.

jasmussen replied:
Ah, gotcha. I don't have strong opinions here, and I would defer to you two for the technical implementation. What matters is:

- We encourage the use of SVG (i.e. make it easy)
- We can provide an icon set even for new blocks, like Dashicons, in the future.
- 3rd party provided SVGs are processed so `role` and `focusable` attributes are added.

If those values are upheld, 👍 👍 from me.

gziolo replied:
> We can provide an icon set even for new blocks, like Dashicons, in the future.

I think it would be convenient to be able to use strings if someone is happy with the default icon set.

> 3rd party provided SVGs are processed so role and focusable attributes are added.

This is what `SVG` component does :)

gziolo commented:
> I don't like that this is forcing us to add a dependency: blocks -> components. Makes me think these primitives should be better in the element package.

It's no longer the case, see https://github.com/WordPress/gutenberg/pull/10652#discussion_r227689332.

> This white-screens for me (and apparently in Travis) with the following error:
> 
> ```
> Uncaught TypeError: Cannot read property 'SVG' of undefined
> ```
> 
> Maybe missing a dependency? (From `blocks` to `components`)

e2e tests pass so I think block icons should be fine as @jorgefilipecosta build an extensive set of assertions for them. It was missing dependency, but I solved it by making `@wordpress/blocks` components agnostic.

gziolo committed:
Ensure that default icon goes through the process of width and height normalization in BlockIcon
Props to @jasmussen for helping to catch it.

gziolo commented:
I added fbd3a30 to ensure that fix for the regression (#10987) works with the default SVG icon, too.

mzorz reviewed (approved):
great work @gziolo - there's a conflict to be resolved in `packages/editor/src/components/block-icon/index.js` but otherwise LGTM after that one gets fixed
